cli/command/image/build/context.go

```diff
@@ -81,59 +81,76 @@ func ValidateContextDirectory(srcPath string, excludes []string) error {
 	})
 }
 
-// GetContextFromReader will read the contents of the given reader as either a
-// Dockerfile or tar archive. Returns a tar archive used as a context and a
-// path to the Dockerfile inside the tar.
-func GetContextFromReader(r io.ReadCloser, dockerfileName string) (out io.ReadCloser, relDockerfile string, err error) {
-	buf := bufio.NewReader(r)
+// DetectArchiveReader detects whether the input stream is an archive or a
+// Dockerfile and returns a buffered version of input, safe to consume in lieu
+// of input. If an archive is detected, isArchive is set to true, and to false
+// otherwise, in which case it is safe to assume input represents the contents
+// of a Dockerfile.
+func DetectArchiveReader(input io.ReadCloser) (rc io.ReadCloser, isArchive bool, err error) {
+	buf := bufio.NewReader(input)
 
 	magic, err := buf.Peek(archiveHeaderSize)
 	if err != nil && err != io.EOF {
-		return nil, "", errors.Errorf("failed to peek context header from STDIN: %v", err)
+		return nil, false, errors.Errorf("failed to peek context header from STDIN: %v", err)
 	}
 
-	if IsArchive(magic) {
-		return ioutils.NewReadCloserWrapper(buf, func() error { return r.Close() }), dockerfileName, nil
-	}
+	return ioutils.NewReadCloserWrapper(buf, func() error { return input.Close() }), IsArchive(magic), nil
+}
 
-	if dockerfileName == "-" {
-		return nil, "", errors.New("build context is not an archive")
+// WriteTempDockerfile writes a Dockerfile stream to a temporary file with a
+// name specified by DefaultDockerfileName and returns the path to the
+// temporary directory containing the Dockerfile.
+func WriteTempDockerfile(rc io.ReadCloser) (string, error) {
+	dockerfileDir, err := ioutil.TempDir("", "docker-build-tempdockerfile-")
+	if err != nil {
+		return "", errors.Errorf("unable to create temporary context directory: %v", err)
 	}
 
-	// Input should be read as a Dockerfile.
-	tmpDir, err := ioutil.TempDir("", "docker-build-context-")
+	f, err := os.Create(filepath.Join(dockerfileDir, DefaultDockerfileName))
 	if err != nil {
-		return nil, "", errors.Errorf("unable to create temporary context directory: %v", err)
+		return "", err
+	}
+	defer f.Close()
+	if _, err := io.Copy(f, rc); err != nil {
+		return "", err
 	}
+	return dockerfileDir, rc.Close()
```

Comment:
If `rc.Close()` returns an error then the callers will leak `tmpDir`, since they don't clean it up in the error case (and it would be a little odd to expect them to do so).

The dir is also leaked on errors in in the `os.Create` above too.

Reply:
@ijc Addressed.